subprojects/core/src/main/java/org/gradle/caching/internal/controller/impl/DefaultBuildCacheCommandFactory.java

```diff
@@ -108,45 +103,30 @@ public OriginMetadata getOriginMetadata() {
                         }
 
                         @Override
-                        public ImmutableSortedMap<String, CurrentFileCollectionFingerprint> getResultingSnapshots() {
+                        public ImmutableSortedMap<String, FileSystemSnapshot> getResultingSnapshots() {
                             return snapshots;
                         }
                     };
                 }
             };
         }
 
-        private ImmutableSortedMap<String, CurrentFileCollectionFingerprint> snapshotUnpackedData(Map<String, ? extends CompleteFileSystemLocationSnapshot> treeSnapshots) {
-            ImmutableSortedMap.Builder<String, CurrentFileCollectionFingerprint> builder = ImmutableSortedMap.naturalOrder();
-            FingerprintingStrategy fingerprintingStrategy = AbsolutePathFingerprintingStrategy.IGNORE_MISSING;
+        private ImmutableSortedMap<String, FileSystemSnapshot> snapshotUnpackedData(Map<String, ? extends CompleteFileSystemLocationSnapshot> treeSnapshots) {
+            ImmutableSortedMap.Builder<String, FileSystemSnapshot> builder = ImmutableSortedMap.naturalOrder();
             entity.visitOutputTrees((treeName, type, root) -> {
                 CompleteFileSystemLocationSnapshot treeSnapshot = treeSnapshots.get(treeName);
-                String internedAbsolutePath = stringInterner.intern(root.getAbsolutePath());
-                List<FileSystemSnapshot> roots = new ArrayList<>();
-
                 if (treeSnapshot == null) {
+                    String internedAbsolutePath = stringInterner.intern(root.getAbsolutePath());
                     MissingFileSnapshot missingFileSnapshot = new MissingFileSnapshot(internedAbsolutePath, AccessType.DIRECT);
                     fileSystemAccess.record(missingFileSnapshot);
-                    builder.put(treeName, fingerprintingStrategy.getEmptyFingerprint());
-                    return;
-                }
-
-                switch (type) {
-                    case FILE:
-                        if (treeSnapshot.getType() != FileType.RegularFile) {
-                            throw new IllegalStateException(String.format("Only a regular file should be produced by unpacking tree '%s', but saw a %s", treeName, treeSnapshot.getType()));
-                        }
-                        roots.add(treeSnapshot);
-                        fileSystemAccess.record(treeSnapshot);
-                        break;
-                    case DIRECTORY:
-                        roots.add(treeSnapshot);
-                        fileSystemAccess.record(treeSnapshot);
-                        break;
-                    default:
-                        throw new AssertionError();
+                    builder.put(treeName, FileSystemSnapshot.EMPTY);
```

Comment:
Shouldn't we add the missing file snapshot to the builder as well as we would do if we find a missing root snapshot? Or are there three different possible output states now: missing, empty and a complete directory snapshot?

There are three states depending on the configuration of the task:
- output property is null
- output property is not null and location does not exist
- output property is not null and location does exit

In `visitOutputTrees`/`visitOutputs` we ignore all output properties which are null, so for the execution engine only the last two options are possible.

Now we need to be consistent what we use as output snapshot when the location is missing. It should be either `MISSING` or `EMPTY`. Currently, we seem to use `EMPTY` here and for overlapping outputs and `MISSING` for non-overlapping outputs.

Reply:
We've discussed this today and the changes are in.